core/node/node_framework/src/service/mod.rs

```diff
@@ -132,17 +156,30 @@ impl ZkStackService {
             resource.stored_resource_wired();
         }
 
+        let mut pre_run = pre_run::PreRun {
+            rt_handle: self.runtime.handle().clone(),
+            resources: self.resources,
+            unstarted_tasks: tasks,
+            join_handles: Vec::new(),
+        };
+
+        if let Some(pre_run_hook) = self.setup_hook.take() {
+            let future = pre_run_hook(&mut pre_run);
+            self.runtime.block_on(future)?;
+        }
+
         // Prepare tasks for running.
         let rt_handle = self.runtime.handle().clone();
-        let join_handles: Vec<_> = tasks
-            .iter_mut()
-            .map(|task| {
-                let task = task.task.take().expect(
-                    "Tasks are created by the node and must be Some prior to calling this method",
-                );
-                rt_handle.spawn(task).fuse()
-            })
-            .collect();
+        let mut join_handles: Vec<_> = pre_run.join_handles;
+        let mut tasks = pre_run.unstarted_tasks;
+
+        for task in &mut tasks {
+            let Some(task) = task.task.take() else {
+                // The task was started during the pre-run.
+                continue;
+            };
```

Comment:
Bikeshedding: You can use `if let` here to have more "linear" control flow.

Reply:
I like the explicit `continue` because it comes with a comment explaining how can it happen and why is it fine, so probably I'll keep it as is.